encoding/codecv7_types.go

```diff
@@ -0,0 +1,499 @@
+func (b *daBatchV7) Blob() *kzg4844.Blob {
+	return b.blob
+}
+
+// BlobBytes returns the blob bytes of the batch.
+func (b *daBatchV7) BlobBytes() []byte {
+	return b.blobBytes
+}
+
+// MarshalJSON implements the custom JSON serialization for daBatchV7.
+// This method is designed to provide prover with batch info in snake_case format.
+func (b *daBatchV7) MarshalJSON() ([]byte, error) {
+	type daBatchV7JSON struct {
+		Version           CodecVersion `json:"version"`
+		BatchIndex        uint64       `json:"batch_index"`
+		BlobVersionedHash string       `json:"blob_versioned_hash"`
+		ParentBatchHash   string       `json:"parent_batch_hash"`
+	}
+
+	return json.Marshal(&daBatchV7JSON{
+		Version:           b.version,
+		BatchIndex:        b.batchIndex,
+		BlobVersionedHash: b.blobVersionedHash.Hex(),
+		ParentBatchHash:   b.parentBatchHash.Hex(),
+	})
+}
+
+// Version returns the version of the DABatch.
+func (b *daBatchV7) Version() CodecVersion {
+	return b.version
+}
+
+// SkippedL1MessageBitmap returns the skipped L1 message bitmap of the DABatch.
+// Note: For daBatchV7, there is no skipped L1 message bitmap, therefore the function returns nil.
+func (b *daBatchV7) SkippedL1MessageBitmap() []byte {
+	return nil
+}
+
+// DataHash returns the data hash of the DABatch.
+// Note: For daBatchV7, there is no data hash, therefore the function returns an empty hash.
+func (b *daBatchV7) DataHash() common.Hash {
+	return common.Hash{}
+}
+
+type blobPayloadV7 struct {
+	prevL1MessageQueueHash common.Hash
+	postL1MessageQueueHash common.Hash
+
+	// used for encoding
+	blocks []*Block
+
+	// used for decoding
+	daBlocks       []DABlock
+	l2Transactions []types.Transactions
+}
+
+func (b *blobPayloadV7) PrevL1MessageQueueHash() common.Hash {
+	return b.prevL1MessageQueueHash
+}
+
+func (b *blobPayloadV7) PostL1MessageQueueHash() common.Hash {
+	return b.postL1MessageQueueHash
+}
+
+func (b *blobPayloadV7) Blocks() []DABlock {
+	return b.daBlocks
+}
+
+func (b *blobPayloadV7) Transactions() []types.Transactions {
+	return b.l2Transactions
+}
+
+func (b *blobPayloadV7) Encode() ([]byte, error) {
+	payloadBytes := make([]byte, blobPayloadV7MinEncodedLength)
+
+	copy(payloadBytes[blobPayloadV7OffsetPrevL1MessageQueue:blobPayloadV7OffsetPostL1MessageQueue], b.prevL1MessageQueueHash[:])
+	copy(payloadBytes[blobPayloadV7OffsetPostL1MessageQueue:blobPayloadV7OffsetInitialL2BlockNumber], b.postL1MessageQueueHash[:])
+
+	var initialL2BlockNumber uint64
+	if len(b.blocks) > 0 {
+		initialL2BlockNumber = b.blocks[0].Header.Number.Uint64()
+		binary.BigEndian.PutUint64(payloadBytes[blobPayloadV7OffsetInitialL2BlockNumber:blobPayloadV7OffsetNumBlocks], initialL2BlockNumber)
+	}
+	binary.BigEndian.PutUint16(payloadBytes[blobPayloadV7OffsetNumBlocks:blobPayloadV7OffsetBlocks], uint16(len(b.blocks)))
+
+	var l1MessageIndex *uint64
+	var transactionBytes []byte
+	for i, block := range b.blocks {
+		if !block.Header.Number.IsUint64() {
```

Comment:
We're a bit inconsistent about these sanity checks, there's a few places where we don't do them.

Reply:
I did some refactoring in 01bd9b5cfff474219eb15ce363937a12847e9451 and now the sanity checks should all reuse the same functions with hopefully all the sanity checks.